Selectively reading sample components with `subset`

cczhu opened the pull request:
Implemented the `subset` property in stream readers, which allows for selectively reading the components (threads, polarisation or channels) of a complete sample.  Any indexing object accepted by a `numpy.ndarray` can be used, except for `None`.  For multi-dimensional indexing, the enclosing structure must be a tuple (i.e. `(dim_1_indexer, dim_2_indexer, ...)`), so multi-dimensional arrays are not accepted.  Advanced indexing that changes the dimensions of the sample shape is also not possible, except for passing single integers since these are transformed into slices within `VLBIStreamBase._get_subset_and_sample_shape`.  Once set, `subset` is used as-is to slice frame data in all formats except VDIF, where threads and channels are sliced separately to retain the selective decoding of frameset frames.  `subset` is a read-only property because it is currently infeasible to re-read a frame every time `subset` changes.

Some notes on design choices:

- As noted above, if the user passes a single number `i` for one dimension, subset's setter turns it into `slice(i, i+1)`. This is because a `namedtuple` with N elements cannot be set with fewer than N objects, and figuring out which dimensions to excise from it (similar to what we do for squeezed shapes in the `sample_shape` getter) requires essentially the same machinery as for converting integers to slices in `subset`.  The latter, however, retains user control of squeezing.
- VDIF reads the first frameset twice now, the first time only to obtain the number of threads. This is ugly, but I'm happy to keep it as is if it turns out not to slow down data read-in very much. Alternatively, can create a `find_nthreads` method in `VDIFFileReader`.
- For consistency with other formats, I'm currently allowing VDIF frames to be returned in the thread ID order the user wishes.  I don't think my implementation of this is self-consistent though, since I still use `return cls(frames, header0)` at the end of `VDIFFrameSet.fromfile`.  Switching that just to `return cls(frames)` gives me a bunch of time offset errors in the test suite I don't fully understand.
- I haven't decided if `_unsliced_shape` should be a tuple or named tuple. Since we use it in a few places (VDIF, GSB) to read new frames, we could initialize `_unsliced_shape` as a genuine sample shape at the same time as `_sample_shape`.

cczhu commented:
I've just updated the PR to address point 3 from above.  This prevents the sample VDIF file from being read properly because of its timestamp offset issue, so I fixed the file's timestamps and made a note of it in `__init__`.  I'm willing to roll back these changes, but feel it makes the code more self-consistent, and prevents us from having a whole bunch of notes in the documentation that one of our sample files is wrong...

mhvk commented:
On the changed file: I'm happy to have the repaired file available, but would like the original one to stay put (with a new name), and with a smaller set of tests that use it. Eventually, we want to have some "repair-on-the-fly" capabilities, in which, e.g., if the seconds in frames are not all the same, all-zero ones are ignored. It will then be good to have examples of problems encountered in the wild.

p.s. Could you separate this out into a different PR? I.e., just make SAMPLE_VDIF the repaired file, add a new SAMPLE_VDIF_... and just check ones that read the streams of both one gets consistent start_time, stop_time and data.

mhvk commented on baseband/vlbi_base/base.py:
```diff
@@ -48,18 +48,19 @@ class VLBIStreamBase(VLBIFileBase):
         self.samples_per_frame = samples_per_frame
         self.sample_rate = sample_rate
         self.offset = 0
+        self._get_subset_and_sample_shape(subset)
```
One might use setters to do this:
```
self.squeeze = squeeze
self.subset = subset
# evalutate lazyproperty
self.sample_shape
```
Without setters
```
self._squeeze = bool(squeeze)
self._subset = subset  # or maybe `self._get_subset(subset)`
self._sample_shape = self._calculate_sample_shape()
```

I think that by looking at both `subset` and `squeeze` in `calculate_sample_shape`, it should be possible to keep integers in subsets for `squeeze=True`.

mhvk commented on baseband/vlbi_base/tests/test_vlbi_base.py:
```diff
@@ -425,34 +425,121 @@ def test_vlbi_file_base(self, tmpdir):
-        assert sbs._unsqueeze(data).shape[1:] == sample_shape_short
+        assert sbs._unsqueeze(data).shape[1:] == unsliced_shape_short
+
+    @pytest.mark.parametrize(('subset', 'sliced_shape',
```
You can use two @parametrize to get the "product" or different ones (or `itertools.product`)

mhvk commented on docs/baseband/tutorials/getting_started.rst:
```diff
@@ -256,25 +256,66 @@ VDIF file's headers are of class::
 
-Opening Specific Threads/Channels From Files
---------------------------------------------
+Opening Specific Components of the Data
```
Note somewhere that it is better to put `slice` than fancy indices in `subset`

mhvk replied:
And put somewhere that `squeeze` is immutable.

mhvk replied:
Maybe for the title of the section, state `Reading Specific Components..` or `Opening a File for Reading Only Specific Components`

cczhu commented:
Addressed the PR comments, and made some revisions to `vlbi_base`.  Now:

- `_get_subset` both processes `subset` and calculates `subset_wrapints`, which converts lone integeres to slices.  The latter is then used in `_get_sample_shape`.  When `squeeze = False`, `subset` is set to `subset_wrapints` to preserve dimensions of length unity.  When `squeeze = True`, we use `subset` as is and squeeze any remaining unity dimensions.  The resulting `sample_shape` corresponds to the dimensions of any payload data indexed with `subset` in either case.
- `_unsliced_shape` is now a namedtuple (of the same type as `sample_shape`).  Stream writers need it to create `_data` and use `_unsqueeze()` (we can make `_data` squeezed, but we'd still need to unsqueeze it when writing to payload).  Moreover, it's needed for the Mark 5B reader because `nchan` is not in the header and `sample_shape` could be subset.  This somewhat lessens the impact of @mhvk's recent changes (since we've just replaced calls to `_sample_shape` with calls to `_unsliced_shape`).
- For VDIF, reverted from using `np.arange(nthread)` to directly obtaining frame numbers from the sorted frameset to set `_thread_ids`.  Moved the machinery to `VDIFStreamReader.__init__`.  This is just in case we encouter VDIF files where the frame number skips values (eg. [0, 1, 629, 630]) but everything else is fine.  Dana's encountered files that skip frame number values, but also have additional problems like invalid frames, and the reader will still break in those cases, but this change should make it easier to hack.  This "feature" is currently untested and undocumented since it's just an implementation choice.
- That said, having a try/except block to handle `self.subset[0]` when setting `_thread_ids` is kind of ugly right now, which is the price we pay for allowing `self.subset[0]` to be an int when `squeeze` is `True`.

mhvk commented on baseband/vlbi_base/tests/test_vlbi_base.py:
```diff
@@ -425,34 +425,147 @@ def test_vlbi_file_base(self, tmpdir):
-        smp_shp_cls = namedtuple('SampleShape',
-                                 'n1, n2, n3, n4, n5, n6, n7, n8')
-        sample_shape = smp_shp_cls(1, 17, 3, 2, 1, 5, 1, 1)
+    def make_streamwriter_with_shape(self, smp_shp_mkr, unsliced_shape,
```
Why not just call the argument `sample_shape_maker`?

cczhu replied:
To stop having to wrap so many lines.  I've expanded the variable as suggested, while shortening the function name.

mhvk replied:
OK, sounds good.

mhvk commented on baseband/vlbi_base/tests/test_vlbi_base.py:
```diff
@@ -425,34 +425,147 @@ def test_vlbi_file_base(self, tmpdir):
+        assert sb.sample_shape._fields == ('n1', 'n2', 'n3', 'n5')
+
+        data = np.empty((100,) + sb.sample_shape, dtype='float32')
+        assert sb._unsqueeze(data).shape[1:] == unsliced_shape
```
Maybe `assert sb._unsqueeze(data).shape == (100,) + unsliced_shape`

mhvk commented on baseband/vlbi_base/tests/test_vlbi_base.py:
```diff
@@ -425,34 +425,147 @@ def test_vlbi_file_base(self, tmpdir):
         assert sbs.sample_shape == ()
         data = np.empty(100, dtype='float32')
-        assert sbs._unsqueeze(data).shape[1:] == sample_shape_short
+        assert sbs._unsqueeze(data).shape[1:] == unsliced_shape_short
```
As above

mhvk commented on baseband/vlbi_base/tests/test_vlbi_base.py:
```diff
@@ -425,34 +425,147 @@ def test_vlbi_file_base(self, tmpdir):
+        # When squeeze = False, passing ints returns slices.
+        sb = self.make_streamwriter_with_shape(smp_shp_mkr, unsliced_shape,
+                                               subset=None, squeeze=False)
+        subset, subset_wrapints = sb._get_subset(0)
```
I would generally avoid testing private functions, as those are implementation details. Just check that you get the right answers instead.

cczhu replied:
Switched to just making new classes and attempting to subset with them.

mhvk commented on docs/baseband/tutorials/getting_started.rst:
```diff
@@ -83,39 +83,37 @@ decoded data array.  Reading files with multiple threads and channels will
-The ``squeeze`` attribute defaults to `False` for all stream readers and
-writers, but this behaviour can be overridden by passing ``squeeze=False`` to
-``open``
+``squeeze`` cannot be changed for an already-open file reader.
```
I don't think this is necessary, as it is generally true that nothing can be changed in a file handle (I know, inconsistent with another comment....)

mhvk commented on docs/baseband/tutorials/getting_started.rst:
```diff
@@ -256,25 +254,71 @@ VDIF file's headers are of class::
+By default, ``fh.read()`` returns complete samples, i.e. with all
+available threads, polarizations or channels. If we were only interested in
+decoding specific components of the complete sample, we can select them by
+passing indexing objects the ``subset`` keyword in open.  For example, if we
```
objects *in* the ``subset``

mhvk commented on docs/baseband/tutorials/getting_started.rst:
```diff
@@ -256,25 +254,71 @@ VDIF file's headers are of class::
+
+Since ``squeeze=False``, ``subset`` is converted from ``3`` to ``slice(3, 4,
+None)`` to retain dimensions of length unity.  This behaviour is turned off
+when ``squeeze=True`` (see below).  Like ``squeeze``, ``subset`` cannot be
```
Again, I feel that we do not have to mention that `squeeze` is immutable.

mhvk commented on docs/baseband/tutorials/getting_started.rst:
```diff
@@ -256,25 +254,71 @@ VDIF file's headers are of class::
+    SampleShape(nthread=2, nchan=1)
+    >>> fh.close()
+
+Here, we have also selected 1 and 3, and channel 0.  No enclosing `tuple` is
```
I don't understand what the text tries to say here. There is only one channel, so we haven't selected anything. I would just delete the sentence.

mhvk commented on docs/baseband/tutorials/getting_started.rst:
```diff
@@ -256,25 +254,71 @@ VDIF file's headers are of class::
+use broadcasting to select specific threads from more than one sample shape
+dimension; see the Numpy documentation on `integer array indexing.
+<https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html#integer-array-indexing>`_
+
```
This is maybe the place to mention that, for the general case, things will be slightly faster with `subset=slice(1, 4, 2)` instead of `subset=[1, 3]` (unfortunately, this is not true for this specific example, though... so maybe just raise an issue to remind us).

mhvk commented on baseband/dada/base.py:
```diff
@@ -282,10 +280,11 @@ def read(self, count=None, out=None):
+            if self.subset:
+                data_slice = (data_slice,) + self.subset
+            out[sample:sample + nsample] = (
+                self._frame[data_slice].squeeze() if self.squeeze else
```
You have two different ways of applying the `subset` -- here, you construct a slice and then use it; in some of the other formats, you slice `data` sequentially. The one here is arguably better, though perhaps less clear.

cczhu replied:
I made the switch to this syntax in the VLBI formats, but had to modify some of the lines for VDIF and Mark 4 (the latter because Mark 4 frame's `__getitem__` forbids indexing).

mhvk replied:
Could you raise an issue about `Mark4Frame` not allowing indexing? That may just be an oversight.

mhvk commented:
On my larger comment, see #127 - let's do that after merging this, since this is very close and very large.

cczhu commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -502,24 +504,38 @@ def test_raw_stream(self, tmpdir):
-                assert np.all(fh_n.read() == fh_r.read())
-                assert abs(fh_n.stop_time - fh_n.time) < 1.*u.ns
-                assert abs(fh_n.stop_time - fh_r.stop_time) < 1.*u.ns
+            fh_n = gsb.open(sh, 'rs', raw=sp, sample_rate=fh_r.sample_rate,
```
the enclosing `with...` statement opens both the sample rawdump files for reading, and two binary files (`sh` and `sp`) for writing.  However, it's silly to reset `sh` and `sp` every time, so rewrote this as you suggested.

mhvk replied:
Ah, yes, that makes sense.

cczhu commented on baseband/mark5b/base.py:
```diff
@@ -338,20 +338,20 @@ class Mark5BStreamWriter(VLBIStreamWriterBase, Mark5BFileWriter):
 
         self._data = np.zeros((self.samples_per_frame,
-                               self._sample_shape.nchan), np.float32)
+                               self._unsliced_shape.nchan), np.float32)
```
I spent probably an hour yesterday experimenting with which I liked better.  I like this because it sets a precedent for formats which don't record the sample shape.

mhvk replied:
OK, let's keep as is then.

cczhu commented on baseband/vlbi_base/base.py:
```diff
@@ -48,58 +48,114 @@ class VLBIStreamBase(VLBIFileBase):
+        for item in subset:
+            try:
+                i = item.__index__()
+            except Exception:
```
It also causes an `AttributeError` for lists, but we can just do `except (AttributeError, TypeError):`.

mhvk replied:
OK, makes sense. With `operator.index`, everything does become a `TypeError`, but I'm happy either way.

cczhu commented on baseband/vdif/base.py:
```diff
@@ -308,15 +311,28 @@ def __init__(self, fh_raw, thread_ids=None, sample_rate=None,
+        # Set _thread_ids.  If subsetting, decode first frameset again.
+        if self.subset:
+            thread_ids = np.array(thread_ids)[self.subset[0]]
+            # In cases where self.subset[0] is a lone int, must create list.
```
On further consideration, there's an annoying edge case where I allow `subset` to be a multi-dimensional numpy array for broadcasting purposes.  After a few design iterations I'm doing the simple thing of squeezing `self.subset[0]` if it's a multi-dimensional array, before using your algorithm.

cczhu commented:
Addressed all issues.  Before we merge I might want to squash the last two commits - I was hoping they'd be independent of one another but I fixed a fatal VDIF bug in the last one that was created in the second-last one.

mhvk commented on baseband/vdif/base.py:
```diff
@@ -308,15 +311,31 @@ def __init__(self, fh_raw, thread_ids=None, sample_rate=None,
+        # Set _thread_ids.  If subsetting, decode first frameset again.
+        if self.subset:
+            # Squeeze in case subset[0] uses broadcasting.
+            subset_0 = (self.subset[0].squeeze()
```
I think this can be simplified by just doing
```
thread_ids = np.array(thread_ids)[subset[0]].squeeze()
```
I'll do it in a quick separate commit.

cczhu replied:
Nope, that squeezes `np.array([1])` into `np.array(1)`.

mhvk commented:
@cczhu - I made some final edits and cleanups, and also rebased (in hindsight, that was a mistake since now it is hard to see what I changed, at least on github). I think we can merge if tests pass.

cczhu commented:
Before merging, would you mind squashing the last two of my commits (and
possibly yours as well)?  The older commit has a bug that the newer one
fixes.

On Tue, Mar 6, 2018 at 11:49 AM, Marten van Kerkwijk <
notifications@github.com> wrote:

> *@mhvk* approved this pull request.
>
> All OK now presumably
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mhvk/baseband/pull/118#pullrequestreview-101625289>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAbCC2MaUdpoXlTAcYgci5B2lQ5Grpg2ks5tbr4tgaJpZM4RxiGB>
> .
>

mhvk commented:
@cczhu - I did squash those commits - there are now only 3, your initial 2 (with reworded titles, to be within 72 char), and a final one combining the rest. Since tests passed, I'll merge!

cczhu commented:
There are titles for commit messages?  Is it like with docstrings where the
first line is 72 char long, but the rest can be in paragraph form?

On Tue, Mar 6, 2018 at 12:14 PM, Marten van Kerkwijk <
notifications@github.com> wrote:

> Merged #118 <https://github.com/mhvk/baseband/pull/118>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mhvk/baseband/pull/118#event-1507029486>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAbCC-i4h668Y8h6AI_D5ev0v4CLinYTks5tbsPfgaJpZM4RxiGB>
> .
>

mhvk commented:
Indeed. When you commit, the text in the editor screen tells you what is expected.
